File: apps/backend/app/schemas/hijri.py
from datetime import datetime
from typing import Literal, Optional

from . import NumPyBaseModel

HijriMethod = Literal["umm_al_qura", "local_hisab", "local_rukyat", "ughc"]


class HijriDateSchema(NumPyBaseModel):
    year: int
    month: int
    day: int


class LocationSchema(NumPyBaseModel):
    lat: float
    lon: float
    timezone: str


class HijriAstronomicalDataSchema(NumPyBaseModel):
    """Data astronomis lengkap, termasuk koordinat referensi (untuk UGHC)."""

    moon_altitude: float
    elongation: float
    moon_age: Optional[float] = None
    is_visible: bool
    lat: Optional[float] = None
    lon: Optional[float] = None
    location_name: Optional[str] = None


class HijriExplanationSchema(NumPyBaseModel):
    method: HijriMethod
    after_sunset: bool
    criteria_used: str
    reasoning: list[str]
    decision: str
    astronomical_data: Optional[HijriAstronomicalDataSchema] = None


class HijriDateResponse(NumPyBaseModel):
    method: HijriMethod
    location: LocationSchema
    hijri_date: HijriDateSchema
    explanation: Optional[HijriExplanationSchema] = None
    generated_at: datetime


class HijriEndMonthResponse(NumPyBaseModel):
    method: HijriMethod
    location: LocationSchema
    generated_at: datetime
    today: HijriDateSchema
    estimated_end_of_month: Optional[HijriDateSchema] = None
    visibility: Optional[HijriAstronomicalDataSchema] = None
    message: Optional[str] = None


class GregorianMonthData(NumPyBaseModel):
    month_id: int
    month_name: str
    total_days: int
    day_1_weekday: int  # 0=Senin (Python weekday)


class GregorianYearResponse(NumPyBaseModel):
    year: int
    months: list[GregorianMonthData]

File: apps/web/types/hijri.ts
export type Method = "umm_al_qura" | "local_hisab" | "local_rukyat" | "ughc";

export interface HijriDate {
  year: number;
  month: number;
  day: number;
}

export interface Location {
  lat: number;
  lon: number;
  timezone: string;
}

export interface HijriAstronomicalData {
  moon_altitude?: number;
  elongation?: number;
  moon_age?: number;
  is_visible?: boolean;
  conjunction_before_sunset?: boolean;
  global_visible?: boolean;
}

export interface HijriExplanation {
  method: Method;
  criteria_used?: string;
  reasoning: string[];
  decision?: string;
  astronomical_data?: HijriAstronomicalData;
}

export interface HijriDateResponse {
  method: Method;
  location: Location;
  hijri_date: HijriDate;
  explanation?: HijriExplanation;
  generated_at: string;
}

export interface HijriEndMonthResponse {
  method: Method;
  location: Location;
  generated_at: string;
  today: HijriDate;
  estimated_end_of_month: HijriDate | null;
  estimated_end_of_month_gregorian?: string | null;
  visibility?: HijriAstronomicalData;
  message?: string;
}
